mlflow/store/runs_artifact_repo.py

```diff
@@ -0,0 +1,96 @@
+from six.moves import urllib
+
+from mlflow.exceptions import MlflowException
+from mlflow.store.artifact_repo import ArtifactRepository
+
+
+class RunsArtifactRepository(ArtifactRepository):
+    """
+    Handles artifacts associated with a Run via URIs of the form
+      `runs:/<run_id>/run-relative/path/to/artifact`.
+    It is a light wrapper that resolves the artifact path to an absolute URI then instantiates
+    and uses the artifact repository for that URI.
+
+    The relative path part of ``artifact_uri`` is expected to be in posixpath format, so Windows
+    users should take special care when constructing the URI.
+    """
+
+    def __init__(self, artifact_uri):
+        from mlflow.tracking.artifact_utils import get_artifact_uri
+        from mlflow.store.artifact_repository_registry import get_artifact_repository
+        (run_id, artifact_path) = RunsArtifactRepository.parse_runs_uri(artifact_uri)
+        uri = get_artifact_uri(run_id, artifact_path)
+        assert urllib.parse.urlparse(uri).scheme != "runs"  # avoid an infinite loop
```

Comment:
Makes sense to prevent infinite loops (I also can't see a valid use case for a run/experiment having a `runs:/...` artifact root directory)

Not for this PR, but we could consider enforcing this at experiment creation time (i.e. error out if a user tries to create an experiment with a runs:/ artifact URI, since child runs under that experiment would inherit the runs:/ artifact URI prefix from their parent experiment)